Parsing and Grammar

nickJones67 opened the pull request:
I think this is the final complete grammar. Also added some semantic rules to clear things up. Still working on some nuances for the semantics and lexical analysis like whether or not to include negative numbers

nickJones67 committed:
Grammar (so far)

Not quite done with the grammar yet but trying to finish it tonight. Let me know how it looks and if I missed anything/f'd up

nickJones67 committed:
Merge remote-tracking branch 'refs/remotes/origin/patch-1'

nickJones67 committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into patch-1

nickJones67 committed:
Update and rename Grammar_v2 to Grammar_v3

JonathanFrias commented:
FYI: You don't ned to do v1 v2 etc. If anything you can rename it to be grammar. Git keeps track of it for you. 

If you click on the file and then click "History", you can see the previous versions.

nickJones67 committed:
Create dumbTimeStamp.c

This creates a timestamp from the system time. Looking at time.h for a better alternative

nickJones67 committed:
Merge remote-tracking branch 'refs/remotes/JonathanFrias/master'

Conflicts:
	Steves-bad-grammar

nickJones67 committed:
Merge branch 'master' of https://github.com/nickJones67/cop4710-sequel

nickJones67 committed:
Merge branch 'refs/heads/patch-1'

nickJones67 committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Grammar_work

nickJones67 committed:
Update Grammar_v3

no expressions. just literals

nickJones67 committed:
FinalGrammar

This should be the final grammar

nickJones67 committed:
Merge remote-tracking branch 'refs/remotes/origin/Grammar_work'

nickJones67 committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Grammar_work

nickJones67 committed:
Merge remote-tracking branch 'refs/remotes/origin/Grammar_work'

nickJones67 committed:
small grammar update

added some necessary tokens to the list

JonathanFrias commented on dumbTimeStamp.c:
```diff
@@ -0,0 +1,55 @@
+// prototypes
+
+
+int main (void) {
```
Is the intent behind this just to prove that we can do time calculations?

nickJones67 replied:
I don't know the correct library for getting the time in the format we need so I made this as a placeholder until I get that squared away. As far as I can tell sys/time.h only returns the number of seconds since 1970 yet this is the library Abbassi suggested I use

SCostello84 replied:
Look up ctime ()
On Nov 20, 2015 6:49 PM, "Nick Jones" notifications@github.com wrote:

> In dumbTimeStamp.c
> https://github.com/JonathanFrias/cop4710-sequel/pull/8#discussion_r45533632
> :
> 
> > +// includes
> > +#include <stdio.h>
> > +#include <stdlib.h>
> > +#include <string.h>
> > +#include <sys/time.h>
> > +
> > +// defines
> > +#define     BUFSIZE    200
> > +
> > +// structs
> > +struct timeval tv;
> > +
> > +// prototypes
> > +
> > +
> > +int main (void) {
> 
> I don't know the correct library for getting the time in the format we
> need so I made this as a placeholder until I get that squared away. As far
> as I can tell sys/time.h only returns the number of seconds since 1970 yet
> this is the library Abbassi suggested I use
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/JonathanFrias/cop4710-sequel/pull/8/files#r45533632.

nickJones67 committed:
Create LexicalAnalyzer.c

lexical analyzer and command prompt user interface for xSQLx engine

nickJones67 committed:
update to timeStamp.c

uses time.h instead

JonathanFrias commented on LexicalAnalyzer.c:
```diff
@@ -0,0 +1,488 @@
+
+// Structs
+typedef struct token {              // token link list node
+    char    type[10];               // can be NUM, ID, KW, SYM
```
You can force the compiler to make this only be a NUM, ID, KW or SYM. See https://github.com/JonathanFrias/cop4710-sequel/blob/master/common.h#L33

JonathanFrias commented on LexicalAnalyzer.c:
```diff
@@ -0,0 +1,488 @@
+    printf("::******************************************************************::\n");
+    printf("::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::\n");
+    printf("                                                                      \n");
+    printf("Welcome to wSQLx!\n");
```
:+1:

JonathanFrias commented on LexicalAnalyzer.c:
```diff
@@ -0,0 +1,488 @@
+    token *head = NULL;
+    token *tmp = NULL;
+
+    while (1) {  // Loop for terminal entries
```
I don't want to leave a ton of feedback that isn't welcome, but this loop appears to be getting complex.

If you want I can pick it apart, but otherwise as long as it constructs a ParseTree struct, I don't mind!

nickJones67 replied:
The loop is long but I know it works at least. I'm not familiar with the parse tree structure that you have outlined as I originally thought you were referring to the LL(1) parsing from Compilers. With this code that exists a list of tokens is generated with a token type, number type, and actual value. Does that fall in line with what you had planned for the parsing. Also I have never user void pointers before so I don't know how they fit in

JonathanFrias replied:
I put int void pointer `void*` because I didn't know what the type should be. Which brings up an interesting point. I guess we have just store everything as text, and then convert the types around after. I don't really see a nice way around it. But you can any pointer in a `void*`, but if you need to change it from `void* Field.value` -> `char* Field.value`, that makes sense.

This is slightly different from the parseTree from compilers, but the idea is that we only fill out _our_ `parseTree` with only leaf nodes from the parse. After parsing is done, we don't need the intermediate steps!

I did very recently create a PR that shows how to construct a parseTree https://github.com/JonathanFrias/cop4710-sequel/blob/store/src/parseTreeHelpers.c#L41. You don't have to use this function specifically, but it can give you an example of what you have to do.

nickJones67 committed:
minor update to Grammar_v3

added DATE to field type along with the associated grammar rules

nickJones67 committed:
Update timeStamp.c

nickJones67 committed:
Update Grammar_v3

nickJones67 committed:
Update Grammar_v3

nickJones67 committed:
Update Grammar_v3

nickJones67 committed:
Update timeStamp.c

nickJones67 committed:
Create sql.l

nickJones67 committed:
Create sql.y

nickJones67 committed:
Update sql.l

nickJones67 committed:
Update sql.y

nickJones67 committed:
create makefile

this is the makefile for the sql.l and sql.y only. The executable is called 'trial'

nickJones67 committed:
Rename makefile to makefile for sql.l and sql.y

nickJones67 commented:
I have updated the sql.l and sql.y files along with a makefile. There is a bug in the sql.l file that I still cannot seem to resolve. With a test file containing "create table this (integer qty);" it will tokenize to:

token: CREATE
token: TABLE
token: ID this
token: ( 
token: INTEGER
token: ID qty 
REJECT

For some reason it will not read the ')' or the ending ';' but rather it will go straight to the parsing and reject for obvious reasons. On the other hand the test file containing "create database this;" works just fine. Any help on this matter would be greatly appreciated

SCostello84 commented:
Shouldn't qty be a token instead of an id?
On Nov 25, 2015 2:53 PM, "Nick Jones" notifications@github.com wrote:

> I have updated the sql.l and sql.y files along with a makefile. There is a
> bug in the sql.l file that I still cannot seem to resolve. With a test file
> containing "create table this (integer qty);" it will tokenize to:
> 
> token: CREATE
> token: TABLE
> token: ID this
> token: (
> token: INTEGER
> token: ID qty
> REJECT
> 
> For some reason it will not read the ')' or the ending ';' but rather it
> will go straight to the parsing and reject for obvious reasons. On the
> other hand the test file containing "create database this;" works just
> fine. Any help on this matter would be greatly appreciated
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/JonathanFrias/cop4710-sequel/pull/8#issuecomment-159715240
> .

nickJones67 committed:
Update sql.y

nickJones67 commented:
They are actually all tokens. ID is just the type of token that will be passed to yyparse() to fit into the grammar

SCostello84 commented:
Ah. I'd then will be passed to the table so we can search for it...got it
On Nov 25, 2015 3:57 PM, "Nick Jones" notifications@github.com wrote:

> They are actually all tokens. ID is just the type of token that will be
> passed to yyparse() to fit into the grammar
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/JonathanFrias/cop4710-sequel/pull/8#issuecomment-159729939
> .

nickJones67 committed:
Update sql.l

nickJones67 committed:
Update sql.y

N00781354 commented:
I think I found it, line 37 in your "sql.l" file:

[a-zA-Z][a-zA-Z0-9_]\* ---> [a-zA-Z][a-zA-Z0-9]*

After I removed the underscore at the end, it created tokens and parsed successfully.
Not sure what an underscore does in terms of Lex definitions, but it seems to make it unhappy.

Also, just a few things:  
1. "sql.y' Line 84: I think the syntax for FIELD_DEF is "ID FIELD_TYPE" not
   the other way around. I just checked and Abbassi's specs: 
   "where field-def in turn has the form field-name(ID) field-type [NOT NULL]"
   It could honestly go both ways, but he might use pre-built SQL queries according to
   his own specs.
2. I think we might have to manually convert user input to all caps, this way we can compare table ID's properly to make sure a create command doesn't define a table that already exists. At the moment I can't think of any other way to compare "Table" to "tABLE" properly other than converting them both to all capital letters. I've got some code that can take care of that for us.
3. Finally, all we need to figure out now is how to export "sql.y" and "sql.l" into the C files we've already built and use their main function. Either that, or import ALL of our C files we've built so far into the main within "sql.y". If we can do that, we can easily return tokens from "sql.l" and call Jonathan's "ParseTree" function within "sql.y". This is what I'm working on currently; let me know if you figure, or have already figured, anything out that works!

Hope you guys have a great Thanksgiving!

> I have updated the sql.l and sql.y files along with a makefile. There is a bug in the sql.l file that I still cannot seem to resolve. With a test file containing "create table this (integer qty);" it will tokenize to:
> 
> token: CREATE
> token: TABLE
> token: ID this
> token: ( 
> token: INTEGER
> token: ID qty 
> REJECT
> 
> For some reason it will not read the ')' or the ending ';' but rather it will go straight to the parsing and reject for obvious reasons. On the other hand the test file containing "create database this;" works just fine. Any help on this matter would be greatly appreciated

nickJones67 committed:
Create sql.c

nickJones67 committed:
Update makefile for sql.l and sql.y

nickJones67 committed:
Update sql.l

nickJones67 committed:
Update sql.y

nickJones67 committed:
Update Grammar_v3

nickJones67 committed:
Update sql.y

nickJones67 commented:
Thanks for testing it man! Yea I just put the underscore in there to give more options for IDs but it is far from necessary. I modified my sql.l to make it function with underscores but if in testing later on we find that it doesn't work then I'll just zap that baby right outta there.

You are right about the ID and FIELD_TYPE being out of order so I have made the necessary changes to the grammar and the sql.y file.

As for the case-sensativity I went to Abbassi's office hours a couple days ago to talk to him about that. He told me that he wants the case preserved for ID's like table names and field names but he should be able to type cReAtE dAtAbAsE aJax; and the create database command will execute with the database name aJax. Case is preserved for the ID names so you can have databases named aJax and AJAX. 

I am almost done incorporating the lex and yacc code into the sql.c file. Feel free to try it out with my makefile and my sql.l, sql.y, and sql.c. To run it, just do 
$./trial
and it will run. Right now I am getting a segmentation fault and I don't know why so feel free to find out. I can't do too much more debugging until I get past that little hitch. 

You peeps have a jolly ole' Thanksgiving too!

JonathanFrias commented:
Case insensitive compares built into `<string.h>`. See http://pubs.opengroup.org/onlinepubs/009695399/functions/strcasecmp.html

For most compares, I imagine we won't need to care about case, as we can just pass through user input.

```
                   _  _
                  | || |__ _ _ __ _ __ _  _
                  | __ / _` | '_ \ '_ \ || |
                  |_||_\__,_| .__/ .__/\_, |
     _____ _              _ |_|  |_|  _|__/   _
    |_   _| |_  __ _ _ _ | |_______ _(_)__ __(_)_ _  __ _
      | | | ' \/ _` | ' \| / (_-< _` | |\ V /| | ' \/ _` |
      |_| |_||_\__,_|_||_|_\_|__|__, |_| \_/ |_|_||_\__, |
                                |___/      .-.      |___/
             .-"-.                       .'  |
        .-_/`     | __                 .'.   ' _
      .-L( o\    ;  -                 /   ` t'` `)
     /.-. \-'   /-.                 .'    ,'\   /  ___
        _\ 7   /-._).--'".-''""--. /        ,-r-.,
       '~| /-. \   (_,-'`         `.          |   )  ___
         /`   \ `._,'                        _,.-' __
               '.                             ;  )
              , ,`.        _.      "-.        f-'
             /`/^. |   .-'`           |    '-/_)
            ; .' | ;  (____,          '|`-. ')
            | |  | ;    .-'          / ;   `~
            '.   |  \  (__..-'     .' /__
             |   \___`-__ (____..-' .'(  '"-..__  ___
              \   r------.._____.' |   `._      `-.  ___
              /  .'  __          `"\     /.---.`-,_)
              |  | __              \\   //     `._)`
              \_.'  -               \\_//         `
                                     `-'
```

N00781354 commented:
Nice work!
For some reason when trying to 'make' I get: "sql.y:129: error: expected identifier or â(â before numeric constant"

It seems to be occurring at "yywrap(scanner)" on line 129. Not sure why though; is this the error you were talking about?

I coped your "sql.y", "sql.l", and "sql.c" files over exactly.. @_@...
Maybe I'm just missing something.. If you're not getting this error I wouldn't worry too much about it, I was just going to try and investigate the segmentation fault.

> I am almost done incorporating the lex and yacc code into the sql.c file. Feel free to try it out with my makefile and my sql.l, sql.y, and sql.c. To run it, just do 
> $./trial
> and it will run. Right now I am getting a segmentation fault and I don't know why so feel free to find out. I can't do too much more debugging until I get past that little hitch.

nickJones67 commented:
Yes! Yea that is the error I was getting before I started getting the segmentation fault. I don't quite understand why that error shows up, although I modified the .l and .y code based on 5 or more different sources so I'm not entirely sure everything is uniformly correct. The scanner I pass to yywrap() is declared the same type that the prototype asks for so I don't know where the problem actually is

On Wed, Nov 25, 2015 at 9:45 PM -0800, "Josiah Hamilton" notifications@github.com wrote:

Nice work!

For some reason when trying to 'make' I get: "sql.y:129: error: expected identifier or â(â before numeric constant"

It seems to be occurring at "yywrap(scanner)" on line 129. Not sure why though; is this the error you were talking about?

I coped your "sql.y", "sql.l", and "sql.c" files over exactly.. @_@...

Maybe I'm just missing something.. If you're not getting this error I wouldn't worry too much about it, I was just going to try and investigate the segmentation fault.

I am almost done incorporating the lex and yacc code into the sql.c file. Feel free to try it out with my makefile and my sql.l, sql.y, and sql.c. To run it, just do 

$./trial

and it will run. Right now I am getting a segmentation fault and I don't know why so feel free to find out. I can't do too much more debugging until I get past that little hitch.

—
Reply to this email directly or view it on GitHub.

nickJones67 committed:
Update common.h

nickJones67 committed:
Update sql.l

nickJones67 committed:
Update sql.y

nickJones67 committed:
Update common.h

nickJones67 committed:
Update common.h

nickJones67 committed:
Update sql.c

nickJones67 commented:
I have updated common.h, sql.l, sql.y, sql.c, and makefile and here is what we have working:
1. yyparse() is called from our main function in sql.c
2. user input is accepted in a while loop until 'exit' is entered
3. lexical analysis and parsing are correct
4. input is not case sensitive (so 'cReaTe DatabAse that;' works just fine)

Working on making the correct calls for creating the command struct in a bit, but in the meantime please test and make sure the above assertions are correct and see if I have overlooked anything.

Commit:
merge code from upstream

nickJones67 committed:
Merge pull request #2 from JonathanFrias/nickJones67-master

Code Sync with Jonathan

nickJones67 committed:
Update common.h

nickJones67 committed:
Update sql.c

nickJones67 committed:
WIP(Work in Progress) commit

nickJones67 committed:
Merge remote-tracking branch 'origin/master'

Conflicts:
	.gitignore
	timeStamp.c

nickJones67 committed:
Do merge

nickJones67 committed:
Merge remote-tracking branch 'refs/remotes/JonathanFrias/master'

nickJones67 committed:
add spaces

nickJones67 committed:
WIP Working on getting flex/bison to work properly

nickJones67 committed:
got stuff to compile

nickJones67 commented:
I committed code that is not quite working yet to my branch but is mostly complete for src/common.h, src/commandHelpers.c, src/commandHelpers.h, and sql.y. I have been debugging all morning but I have to go to work to if someone could take a crack at it in the meantime it would be greatly appreciated.

nickJones67 committed:
working commands: create database, load database, save; in progress: create table

nickJones67 committed:
message

nickJones67 committed:
terminal with proper memory allocation

nickJones67 committed:
Somewhat working and it does stuff

nickJones67 committed:
all the running changes to the lex and yacc

nickJones67 committed:
Got select, insert kinda working.

Still issues with statements with multiple fields

nickJones67 committed:
Things are looking better kind of. Here's what works: create table, drop table, insert into (with multiple fields), create database, drop database, select all

nickJones67 committed:
adding a small tester file

nickJones67 committed:
Change to tester with new add field func
